p4-16/spec/P4-16-spec.mdk

```diff
@@ -4180,11 +4180,38 @@ h = { 10, 12 };  // This also makes the header h valid
 h = { y = 12, x = 10 };  // Same effect as above
 ~ End P4Example
 
-Two headers can be compared for equality (==) or inequality (!=) only
+Two headers can be compared for equality (`==`) or inequality (`!=`) only
 if they have the same type.  Two headers are equal if and only if they
 are both invalid, or they are both valid and all their corresponding
 fields are equal.
 
+The expression `{#}` represents an invalid header of some type, but it
```

Comment:
represents an invalid header or an invalid header union of an unspecified type.
The expression `(H){#}` represents an invalid header or invalid header union with a specified type.

Reply:
So this is a minor detail, perhaps, but if I say the following things in the spec:

+ The expression `{#}` represents an invalid header or an invalid header union of an unspecified type.
+ If the type of `{#}` cannot be inferred from the context, the program will be rejected.
+ You may use a cast to convert the type of `{#}` to that of the header or header union type specified by the cast.

Then it seems to me that saying "The expression `(H) {#}` represents an invalid header or invalid header union with a specified type." is on a slippery slope to what I think is one of Nate's objections, which is that it sounds like you are specifying a NEW KIND of expression, which it really is not.  It is simply `{#}` with a cast on it.

Reply:
Actually inside the compiler we have three kinds of expressions; these are defined in expressions.def:
`{#}` class Invalid
`(H)` class Cast
`(H){#}` class InvalidHeader or InvalidHeaderUnion
So the description is accurate.
It is just that the parser produces `(cast)(Invalid)` and the typechecker converts that to `InvalidHeader`.

Reply:
If you and Nate can come to agreement on what wording you think belongs in the spec for this, and hopefully the same for the expression that represents a value that is an invalid header union, then I suspect most other people would be willing to accept that.

I've got two PRs with my best guesses, but I don't know which of the comments on this PR represent suggested changes to this PR, and which of those Nate would also approve of.